### PR TITLE
Pbi/91606 uri token offers

### PR DESCRIPTION
**Pull request by kasunkavinda**

### CODE REVIEW

File: src/clients/host-client.js [hunk 1]
```diff
@@ -45,24 +45,15 @@ class HostClient extends BaseEvernodeClient {
         super(xrpAddress, xrpSecret, Object.values(HostEvents), true, options);
     }
 
-    async getRegistrationNft() {
+    async getRegistrationUriTokens() {
```
Comment (chalith) [inline src/clients/host-client.js:48]: getRegistrationUriTokens -> getRegistrationUriToken

File: src/clients/host-client.js [hunk 2]
```diff
@@ -166,13 +156,15 @@ class HostClient extends BaseEvernodeClient {
         // Check whether is there any missed NFT sell offer that needs to be accepted
         // from the client-side in order to complete the registration.
         const registryAcc = new XrplAccount(this.config.registryAddress, null, { xrplApi: this.xrplApi });
-        const regNft = await this.getRegistrationNft();
-        if (!regNft) {
+        const regUriTokens = await this.getRegistrationUriTokens();
```
Comment (chalith) [inline src/clients/host-client.js:159]: regUriTokens -> regUriToken

File: src/clients/host-client.js [hunk 3]
```diff
@@ -166,13 +156,15 @@ class HostClient extends BaseEvernodeClient {
...
+        
+        if (!regUriTokens) {
             const regInfo = await this.getHostInfo(this.xrplAcc.address);
             if (regInfo) {
-                const sellOffer = (await registryAcc.getNftOffers()).find(o => o.NFTokenID == regInfo.nfTokenId);
+                const sellOffer = (await registryAcc.getURITokens()).find(o => o.index == regInfo.nfTokenId);
```
Comment (chalith) [inline src/clients/host-client.js:164]: Shall we change the condition to `o.index == regInfo.nfTokenId && 0.Amount` to verify whether there's an offer

File: src/clients/host-client.js [hunk 4]
```diff
@@ -166,13 +156,15 @@ class HostClient extends BaseEvernodeClient {
...
+                console.log('sell offer')
                 if (sellOffer) {
-                    await this.xrplAcc.buyNft(sellOffer.index);
+                    await this.xrplAcc.buyURIToken(sellOffer.index, XrplConstants.MIN_XRP_AMOUNT, XrplConstants.XRP);
```
Comment (chalith) [inline src/clients/host-client.js:167]: What if we change `buyURIToken(uriTokenID, amount, currency, issuer = null, options = {})` to `buyURIToken(uriToken, issuer = null, options = {})`?
I think then we can directly use `uriToken.Amount` for Amount in `URITokenBuy` transaction

File: src/clients/host-client.js [hunk 5]
```diff
@@ -198,7 +190,7 @@ class HostClient extends BaseEvernodeClient {
         catch (e) {
             console.log("No initiated transfers were found.");
         }
-
+
```
Comment (chalith) [inline src/clients/host-client.js:193]: CTRL + Shift + I

File: src/clients/host-client.js [hunk 6]
```diff
@@ -219,14 +211,16 @@ class HostClient extends BaseEvernodeClient {
...
-        console.log('Waiting for the sell offer')
+        console.log('Waiting for the sell offer', tx)
         let offer = null;
         let attempts = 0;
         let offerLedgerIndex = 0;
         while (attempts < OFFER_WAIT_TIMEOUT) {
-            const nft = (await registryAcc.getNfts()).find(n => (n.URI === `${EvernodeConstants.NFT_PREFIX_HEX}${tx.id}`) || (n.NFTokenID === transferredNFTokenId));
-            if (nft) {
-                offer = (await registryAcc.getNftOffers()).find(o => o.Destination === this.xrplAcc.address && o.NFTokenID === nft.NFTokenID && o.Flags === 1);
+            const uriTokens = (await registryAcc.getURITokens()).find(n => (n.URI === `${EvernodeConstants.NFT_PREFIX_HEX}${tx.id}`) || (n.index === transferredNFTokenId));
```
Comment (chalith) [inline src/clients/host-client.js:219]: Can't we combine these two finds.
```
const uriTokens = (await registryAcc.getURITokens()).find(n => ((n.URI === `${EvernodeConstants.NFT_PREFIX_HEX}${tx.id}`) || (n.index === transferredNFTokenId)) && n.Amount && o.Destination === this.xrplAcc.address);
```

File: src/clients/host-client.js [hunk 7]
```diff
@@ -339,7 +339,7 @@ class HostClient extends BaseEvernodeClient {
 
         // To obtain registration NFT Page Keylet and index.
         if (!tokenID)
-            tokenID = (await this.getRegistrationNft()).NFTokenID;
+            tokenID = (await this.getRegistrationUriTokens()).NFTokenID;
         const nftPageDataBuf = await EvernodeHelpers.getNFTPageAndLocation(tokenID, this.xrplAcc, this.xrplApi);
```
Comment (chalith) [inline src/clients/host-client.js:343]: Add a `TODO: Need to replace with URI token keylet` comment in places where `getNFTPageAndLocation` used for future reference

File: src/clients/host-client.js [hunk 8]
```diff
@@ -45,24 +45,15 @@ class HostClient extends BaseEvernodeClient {
...
-            // Check whether the token was actually issued from Evernode registry contract.
-            const issuerHex = nft.NFTokenID.substr(8, 40);
-            const issuerAddr = codec.encodeAccountID(Buffer.from(issuerHex, 'hex'));
-            if (issuerAddr == this.config.registryAddress) {
-                return nft;
-            }
-        }
-
-        return null;
+        const nft = (await this.xrplAcc.getURITokens()).find(n => n.index.startsWith(EvernodeConstants.NFT_PREFIX_HEX) && n.Issuer === this.config.registryAddress);
```
Comment (Udith-Gayan) [inline src/clients/host-client.js:50]: should be n.URI.startsWith.....

File: src/clients/host-client.js [hunk 9]
```diff
@@ -45,24 +45,15 @@ class HostClient extends BaseEvernodeClient {
...
+        return nft ?? null;
     }
 
     async getRegistration() {
         // Check whether we own an evernode host token.
-        const nft = await this.getRegistrationNft();
+        const nft = await this.getRegistrationUriToken();
         if (nft) {
             const host = await this.getHostInfo();
             return (host?.nfTokenId == nft.NFTokenID) ? host : null;
```
Comment (Udith-Gayan) [inline src/clients/host-client.js:59]: should be     ==  nft.index

File: src/clients/host-client.js [hunk 10]
```diff
@@ -45,24 +45,15 @@ class HostClient extends BaseEvernodeClient {
...
-
-        return null;
+        const nft = (await this.xrplAcc.getURITokens()).find(n => n.index.startsWith(EvernodeConstants.NFT_PREFIX_HEX) && n.Issuer === this.config.registryAddress);
+        return nft ?? null;
     }
 
     async getRegistration() {
         // Check whether we own an evernode host token.
-        const nft = await this.getRegistrationNft();
+        const nft = await this.getRegistrationUriToken();
```
Comment (Udith-Gayan) [inline src/clients/host-client.js:56]: Better we can rename the variable to uriToken

File: src/clients/host-client.js (hunk 8 above)
Comment (Udith-Gayan) [reply src/clients/host-client.js:50]: Better we rename the variable nft  to   uriToken

File: src/xrpl-account.js [hunk 11]
```diff
@@ -582,16 +582,24 @@ class XrplAccount {
...
             tx.Destination = toAddr;
-        
+
         return this.#submitAndVerifyTransaction(tx, options);
     }
 
-    buyURIToken(uriTokenID, amount, currency, issuer = null, options = {}) {
-        const amountObj = makeAmountObject(amount, currency, issuer);
+    buyURIToken(uriToken, issuer = null, options = {}) {
+        const amountObj = makeAmountObject(uriToken.Amount, EvernodeConstants.EVR, issuer);
```
Comment (Udith-Gayan) [inline src/xrpl-account.js:590]: Currency type is not always EVR.

File: src/clients/host-client.js [hunk 12]
```diff
@@ -117,8 +108,7 @@ class HostClient extends BaseEvernodeClient {
         const uri = uriBuf.toString('hex').toUpperCase();
 
         await this.xrplAcc.mintURIToken(uri, null, { isBurnable: true, isHexUri: true });
-
-        const uriToken  = await this.xrplAcc.getURITokenByUri(uri, true);
+        const uriToken = await this.xrplAcc.getURITokenByUri(uri, true);
         if (!uriToken)
             throw "Offer lease NFT creation error.";
```
Comment (Udith-Gayan) [inline src/clients/host-client.js:114]: In 115 line, should be changed to sellURIToken()

**Review by kithminisg — APPROVED** [review]

**Pull request merged**